base/range.jl

```diff
@@ -737,44 +689,43 @@ end
 ## linear operations on ranges ##
 
 -(r::OrdinalRange) = range(-first(r), -step(r), length(r))
--(r::FloatRange)   = FloatRange(-r.start, -r.step, r.len, r.divisor)
--(r::LinSpace)     = LinSpace(-r.start, -r.stop, r.len, r.divisor)
+-(r::StepRangeLen) = StepRangeLen(-r.ref, -r.step, length(r), r.offset)
+-(r::LinSpace) = LinSpace(-r.start, -r.stop, length(r))
 
 +(x::Real, r::AbstractUnitRange) = range(x + first(r), length(r))
-+(x::Real, r::Range) = (x+first(r)):step(r):(x+last(r))
-#+(x::Real, r::StepRange)  = range(x + r.start, r.step, length(r))
-+(x::Real, r::FloatRange) = FloatRange(r.divisor*x + r.start, r.step, r.len, r.divisor)
-function +{T}(x::Real, r::LinSpace{T})
-    x2 = x * r.divisor / (r.len - 1)
-    LinSpace(x2 + r.start, x2 + r.stop, r.len, r.divisor)
-end
-+(r::Range, x::Real)      = x + r
-#+(r::FloatRange, x::Real) = x + r
-
--(x::Real, r::Range)      = (x-first(r)):-step(r):(x-last(r))
--(x::Real, r::FloatRange) = FloatRange(r.divisor*x - r.start, -r.step, r.len, r.divisor)
-function -(x::Real, r::LinSpace)
-    x2 = x * r.divisor / (r.len - 1)
-    LinSpace(x2 - r.start, x2 - r.stop, r.len, r.divisor)
-end
--(r::AbstractUnitRange, x::Real) = range(first(r)-x, length(r))
--(r::StepRange , x::Real) = range(r.start-x, r.step, length(r))
--(r::FloatRange, x::Real) = FloatRange(r.start - r.divisor*x, r.step, r.len, r.divisor)
-function -(r::LinSpace, x::Real)
-    x2 = x * r.divisor / (r.len - 1)
-    LinSpace(r.start - x2, r.stop - x2, r.len, r.divisor)
-end
-
-*(x::Real, r::OrdinalRange) = range(x*first(r), x*step(r), length(r))
-*(x::Real, r::FloatRange)   = FloatRange(x*r.start, x*r.step, r.len, r.divisor)
-*(x::Real, r::LinSpace)     = LinSpace(x * r.start, x * r.stop, r.len, r.divisor)
-*(r::Range, x::Real)        = x * r
-*(r::FloatRange, x::Real)   = x * r
-*(r::LinSpace, x::Real)     = x * r
-
-/(r::OrdinalRange, x::Real) = range(first(r)/x, step(r)/x, length(r))
-/(r::FloatRange, x::Real)   = FloatRange(r.start/x, r.step/x, r.len, r.divisor)
-/(r::LinSpace, x::Real)     = LinSpace(r.start / x, r.stop / x, r.len, r.divisor)
+# For #18336 we need to prevent promotion of the step type:
++(x::Number, r::AbstractUnitRange) = range(x + first(r), step(r), length(r))
```

Comment:
I'm almost certain there's a test something like `(1:3)+im`, although I don't know whether the output type is tested.

But yes, there's no particular reason to restrict to `Real` anymore.